streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java

```diff
@@ -84,26 +89,29 @@ private void initInternal(final ProcessorContext context) {
     private void putAndMaybeForward(final ThreadCache.DirtyEntry entry,
                                     final InternalProcessorContext context) {
         if (flushListener != null) {
-            final byte[] newValueBytes = entry.newValue();
-            final byte[] oldValueBytes = newValueBytes == null || sendOldValues ? wrapped().get(entry.key()) : null;
+            final byte[] rawNewValue = entry.newValue();
+            final byte[] rawOldValue = rawNewValue == null || sendOldValues ? wrapped().get(entry.key()) : null;
 
             // this is an optimization: if this key did not exist in underlying store and also not in the cache,
             // we can skip flushing to downstream as well as writing to underlying store
-            if (newValueBytes != null || oldValueBytes != null) {
-                final K key = serdes.keyFrom(entry.key().get());
-                final V newValue = newValueBytes != null ? serdes.valueFrom(newValueBytes) : null;
-                final V oldValue = sendOldValues && oldValueBytes != null ? serdes.valueFrom(oldValueBytes) : null;
+            if (rawNewValue != null || rawOldValue != null) {
+                final FlushEntry<K, V> flushEntry = flushEntry(
+                    serdes,
+                    rawNewValue,
+                    sendOldValues ? rawOldValue : null,
+                    entry.entry().context().timestamp());
```

Comment:
I'm sorry, but I still don't understand the motivation for constructing these objects. Do you mind explaining again?

Reply:
Because `CachingTimestampKeyValueStore` need to overwrite `flushEntry()` (cf https://github.com/apache/kafka/pull/6152/files#diff-25a648e31db0d5090ce7c5c5d247cf71R38)

Reply:
I think @mjsax 's motivation is to centralize the logic for 1) deserializing old and new values, and 2) choosing the timestamp, upon flushing.

Personally I felt it is a bit overkill since we need to create a new class, and also creating a temporary object every time for holding these three values before flushing: take an example of `CachingTimestampedKeyValueStore` line 40, the fact that we actually do not take the passed-in `timestamp` field at all but only need to return the extracted `valueTimestamp` for flushing makes me feeling that this class is not necessary.

I've also reviewed the other PR (https://github.com/apache/kafka/pull/6331) for deferring the deserialization from the caching store all together, and do that in flush listener only, but it is not very elegant either, and after tried several alternatives myself I felt probably keeping the deserialization logic in caching store for now is the "least bad option" we have at hand.

So I'd like to propose a final time to not using a FlushEntry class, but just inlining the logic of deserializing the old/new values, as well as choosing the timestamp to flush within each class, which is:

1) deserializing logic of new / old values for KeyValue, Window, Session stores are the same.
2) deserializing logic of new / old values for TimestampedKeyValue would be different (and potentially this logic can be centralized into a place when we have TimestampedWindow later).

As for which timestamp to flush, I actually do not know why we should use the deserialized timestamp from the cache rather than still use the timestamp field in the cache to flush? My understanding is that they should be the same right? In that case the logic of selecting the timestamp should all be the same.

Reply:
I prefer to optimize for sharing code and following OO principles. However, I want to point out, that I refactored #6331 and hope this new version helps to resolve the issue. If we can agree that moving the deserialization out of the caching layer is the right way, than this flushing logic will become quite simplified.

> As for which timestamp to flush, I actually do not know why we should use the deserialized timestamp from the cache rather than still use the timestamp field in the cache to flush? My understanding is that they should be the same right? In that case the logic of selecting the timestamp should all be the same.

As commented here, they are not the same and we need to get the timestamp from the value; using the timestamp from the LRUCacheEntry would be incorrect: https://github.com/apache/kafka/pull/6152#discussion_r260612137